nomiclabs/hardhat
# Run tests twice

This is a mocha caching and cleanup issue addressed here: mochajs/mocha#2783.

Relates to #1720

## Testing

An integration test has been added to e2e to cover the run twice scenario and guard against regression of the bug.

### Manual

To test manually:

1. create a new advanced typescript hardhat project
2. link to `hardhat-core
3. create a new scripts file `./scripts/multi-run.js`:

```javascript
// eslint-disable-next-line import/no-extraneous-dependencies
const hre = require("hardhat");

async function main() {
  const code = await hre.run("test");

  if (code > 0) {
    console.error("Failed first test run");
    process.exit(1);
  }

  const secondCode = await hre.run("test");

  if (secondCode > 0) {
    console.error("Failed second test run");
    process.exit(1);
  }

  console.log("\nFinished and complete\n");
}

main()
  .then(() => process.exit(0))
  .catch((error) => {
    console.error(error);
    process.exit(1);
  });

```

4. Run the script `npx hardhat run ./scripts/multi-run.js`
5. Confirm the script runs without error and that tests are present in both runs:

![image](https://user-images.githubusercontent.com/24030/145562547-893d522d-ce35-47d0-af48-25c1d0533f75.png)


As a test against a more in-depth test suite, I did the equivalent of the above on the open-zepplin test suite (updating to the latest hardhat libs first) and was able to run the suite twice programmatically with the script above.

## TODO

* [x] Apply fix from mocha suggestions
* [x] Look at options on unit tests
* [x] Run against existing project with large test suite
